File: cracking_the_coding_interview_6th/chapter17/cpp/17_14_v02.cpp
#include <fstream>
#include <queue>
#include <string>
#include <vector>
using namespace std;
using ll = long long;

vector<ll> find_smallest_kth_nums(vector<ll> v, const ll k) {
  if (v.size() <= k) return v;
  priority_queue<ll> q;
  for (ll i = 0; i < k; ++i) {
    q.push(v[i]);
  }
  for (ll i = k; i < v.size(); ++i) {
    ll u = q.top();
    if (u > v[i]) {
      q.pop();
      q.push(v[i]);
    }
  }
  vector<ll> res(k);
  for (ll i = 0; i < k; ++i) {
    res[i] = q.top();
    q.pop();
  }
  return res;
}

void solve(const string &fp) {
  printf("# %s\n", fp.c_str());
  ifstream ifs{fp};
  ll n, k;
  ifs >> n >> k;
  vector<ll> v(n);
  for (ll i = 0; i < n; ++i) {
    ifs >> v[i];
  }
  vector<ll> res = find_smallest_kth_nums(v, k);
  for (ll i = 0; i < res.size(); ++i) {
    printf("%lld%c", res[i], i == res.size() - 1 ? '\n' : ' ');
  }
  puts("");
}

int main() {
  solve("../testcases/17_14/01.txt");
  solve("../testcases/17_14/02.txt");
  solve("../testcases/17_14/03.txt");
}

File: cracking_the_coding_interview_6th/chapter17/python/17_14_v02.py
from typing import List
from heapq import heapify, heappushpop


def find_smallest_kth_nums(v: List[int], k: int) -> List[int]:
    if len(v) <= k:
        return v
    v = list(map(lambda x: -x, v))
    q = v[:k]
    heapify(q)
    for _v in v[k:]:
        if _v > q[0]:
            heappushpop(q, _v)
    return list(map(lambda x: -x, q))


def ns(f):
    return next(f).strip()


def solve(fp: str):
    print(f"# {fp}")
    with open(fp) as f:
        n, k = map(int, ns(f).split())
        v = list(map(int, ns(f).split()))
    res = find_smallest_kth_nums(v, k)
    print(sorted(res))
    print()


solve("../testcases/17_14/01.txt")
solve("../testcases/17_14/02.txt")
solve("../testcases/17_14/03.txt")
